### DISCUSSION TITLE
Modification of the Ajax component so that the user can return a JSON object.

### DISCUSSION THREAD

**Discussion by korenevskiy** [joomla/joomla-cms#43471] [General] [answered]
```php
} elseif (\is_scalar($results)) {
            // Output string/ null
            $out = (string) $results;
} else {
            // Output array/ object
            $out = implode((array) $results);
}
```
[/components/com_ajax/ajax.php](https://github.com/joomla/joomla-cms/blob/076864d32fc529007cf05d003426f24e4cc239f3/)
 
In the com_ajax component, the returned `$results` value is reduced to a string.
Can the user return a `new JsonResponse `type object. 
But then this object will  be converted to an array. and it will be glued together and cause an error.

Do I need to do PR or is this normal behavior? Should we make the code more flexible?

I suggest adding a check to see if the object is a JSON type or not.
```php
} elseif (\is_scalar($results) ||  $results instanceof JsonResponse) {
            // Output string/ null
            $out = (string) $results;
} else {
```

components/com_ajax/ajax.php)

**Comment by Fedik** [comment]
Something like this is possible since 5.0,
you have to use url with `&format=json`.

And the event listener, kind of:
```php
public function onAjaxFooBar(Joomla\CMS\Event\Plugin\AjaxEvent $event)
{
    $response = ['foo' => 'bar', 'bar' => 'foo'];

    $event->updateEventResult($response);
}
```
It set the array, which will be passed to  `JsonResponse` within the `com_ajax`

**Comment by korenevskiy** [comment]
I know that you need to install the JSON format for this. But I needed to pass the format as DEBUG, but at the same time so that the format was in JSON.

The problem is that the object that generates the `new JsonResponse` with the second attribute accepts the error message and must also generate a list of other messages sent
```php
JFactory::getApplication()->enqueueMessage('Error Message');
```
so that later the JSON contains the service information of error messages.
But the Ajax component disables these messages and passes NULL there.

So I need to pass JSON from component c with messages. but the component does not transmit them, but calls JSON to form itself.
If I generate JSON using a string manually, then this will accompany confusion and unnecessary magic.

thus, we need to achieve a JSON mode with error messages. And to do this, you need to add support for this feature, as I indicated in the main description.

**Reply by Fedik** [reply]
To collect messages you can add `&ignoreMessages=0&format=json`, then it will not ignore App messages.

**Reply by korenevskiy** [reply]
> To collect messages you can add `&ignoreMessages=0&format=json`, then it will not ignore App messages.

Before creating a discussion here, I did as you wrote. But thanks for your PR. The tambourine is no longer needed.

**Comment by Fedik** [comment, marked as answer]
I have made PR for it, please test https://github.com/joomla/joomla-cms/pull/43530
